src/op.rs

```diff
@@ -0,0 +1,1288 @@
+                register: regno,
+                offset: offset,
+            }))
+        }
+        constants::DW_OP_piece => {
+            let (newbytes, size) = try!(parse_unsigned_lebe(bytes));
+            Ok((newbytes,
+                Operation::Piece {
+                size_in_bits: 8 * size,
+                bit_offset: None,
+            }))
+        }
+        constants::DW_OP_deref_size => {
+            let (newbytes, size) = try!(parse_u8e(bytes));
+            Ok((newbytes,
+                Operation::Deref {
+                size: size,
+                space: false,
+            }))
+        }
+        constants::DW_OP_xderef_size => {
+            let (newbytes, size) = try!(parse_u8e(bytes));
+            Ok((newbytes,
+                Operation::Deref {
+                size: size,
+                space: true,
+            }))
+        }
+        constants::DW_OP_nop => Ok((bytes, Operation::Nop)),
+        constants::DW_OP_push_object_address => Ok((bytes, Operation::PushObjectAddress)),
+        constants::DW_OP_call2 => {
+            let (newbytes, value) = try!(parse_u16(bytes));
+            Ok((newbytes,
+                Operation::Call { offset: DIEReference::UnitRef(UnitOffset(value as u64)) }))
+        }
+        constants::DW_OP_call4 => {
+            let (newbytes, value) = try!(parse_u32(bytes));
+            Ok((newbytes,
+                Operation::Call { offset: DIEReference::UnitRef(UnitOffset(value as u64)) }))
+        }
+        constants::DW_OP_call_ref => {
+            let (newbytes, value) = try!(parse_word(bytes, format));
+            Ok((newbytes,
+                Operation::Call { offset: DIEReference::DebugInfoRef(DebugInfoOffset(value)) }))
+        }
+        constants::DW_OP_form_tls_address => Ok((bytes, Operation::TLS)),
+        constants::DW_OP_call_frame_cfa => Ok((bytes, Operation::CallFrameCFA)),
+        constants::DW_OP_bit_piece => {
+            let (newbytes, size) = try!(parse_unsigned_lebe(bytes));
+            let (newbytes, offset) = try!(parse_unsigned_lebe(newbytes));
+            Ok((newbytes,
+                Operation::Piece {
+                size_in_bits: size,
+                bit_offset: Some(offset),
+            }))
+        }
+        constants::DW_OP_implicit_value => {
+            let (newbytes, size) = try!(parse_unsigned_lebe(bytes));
+            let size = size as usize;
+            Ok((newbytes, Operation::ImplicitValue { data: newbytes.range(0..size).into() }))
+        }
+        constants::DW_OP_stack_value => Ok((bytes, Operation::StackValue)),
+
+        _ => return Err(Error::InvalidExpression(name)),
+    }
+}
+
+/// A DWARF expression evaluation.
+pub struct Evaluation<'input, Endian>
+    where Endian: Endianity
+{
+    bytecode: &'input [u8],
+    address_size: u8,
+    format: Format,
+    initial_value: Option<u64>,
+    object_address: Option<u64>,
+    callbacks: &'input mut EvaluationContext<'input>,
```

Comment:
This will automatically shorten `'input` to the bounds of the `EvaluationContext`'s lifetime. For maximum flexibility, its best to separate the lifetime of the reference from the generic lifetime parameter out like this:

```rust
pub struct Evaluation<'context, 'input, Endian>
    where Endian: 'context + Endianity,
          'input: 'context
{
    ...
}
```

I don't think this is really problematic in practice in this case because we aren't *returning* any string or slice with the `'input` lifetime and obviously `Evaluation` is bounded by both the reference to the context and the generic lifetime parameter. However, if we were returning a string or slice, we would definitely not want to accidentally shorten its lifetime, or it would be a pain for our users (we did this once already).

All that said, I think we should make this change even if it isn't problematic in this case, just for consistency with the other places where we have this kind of situation.